axiom/nr_segment_message.c

```diff
@@ -0,0 +1,231 @@
+ * be replaced with 'Temp'. The DestinationType segment SHOULD NOT contain
+ * "Temporary". Thus, "Temporary " should be removed from the destination type
+ * enum before metric use. Examples: MessageBroker/JMS/Queue/Produce/Temp,
+ * MessageBroker/JMS/Topic/Produce/Temp
+ *
+ * Further note that for pull-style messaging, the transaction segment name MUST
+ * be equal to the scoped metric name (e.g.,
+ * MessageBroker/JMS/Queue/Produce/Named/SortQueue)
+ *
+ *
+ * Params  : 1. The message segment.
+ *           2. Message parameters
+ *           3. Duration of the segment
+ *
+ * Returns : the scoped metric that was created.  Caller is responsible for
+ * freeing this value.
+ */
+
+static char* nr_segment_message_create_metrics(
+    nr_segment_t* segment,
+    const nr_segment_message_params_t* message_params,
+    nrtime_t duration) {
+  const char* action_string = NULL;
+  const char* destination_type_string = NULL;
+  const char* library_string = NULL;
+  char* rollup_metric = NULL;
+  char* scoped_metric = NULL;
+
+  if (NULL == segment) {
+    return NULL;
+  }
+
+  if (NULL == message_params) {
+    return NULL;
+  }
+
+  /* Rollup metric.
+   *
+   * This has to be created on the transaction in order to create
+   * MessageBroker/allWeb and MessageBroker/allOther and to calculate
+   * messageDuration later on.
+   */
+
+  nrm_force_add(segment->txn->unscoped_metrics, "MessageBroker/all", duration);
+
+  if (nr_strempty(message_params->library)) {
+    library_string = "<unknown>";
+  } else {
+    library_string = message_params->library;
+  }
+  rollup_metric = nr_formatf("MessageBroker/%s/all", library_string);
+  nrm_force_add(segment->txn->unscoped_metrics, rollup_metric, duration);
+  nr_free(rollup_metric);
+
+  /*
+   * Note: although the concept of Temporary queues/topics is detailed in the
+   * spec, in practice, we are unlikely to encounter it as it is currently only
+   * meaningful with JMS (Java Message Service).  It is added here for adherence
+   * with spec.
+   */
+
+  if (NR_SPAN_PRODUCER == message_params->message_action) {
+    action_string = "Produce";
+  } else if (NR_SPAN_CONSUMER == message_params->message_action) {
+    action_string = "Consume";
+  } else {
+    action_string = "<unknown>";
+  }
+
+  switch (message_params->destination_type) {
+    case NR_MESSAGE_DESTINATION_TYPE_TEMP_QUEUE:
+    case NR_MESSAGE_DESTINATION_TYPE_QUEUE:
+      destination_type_string = "Queue";
+      break;
+    case NR_MESSAGE_DESTINATION_TYPE_TEMP_TOPIC:
+    case NR_MESSAGE_DESTINATION_TYPE_TOPIC:
+      destination_type_string = "Topic";
+      break;
+    case NR_MESSAGE_DESTINATION_TYPE_EXCHANGE:
+      destination_type_string = "Exchange";
+      break;
+    default:
+      destination_type_string = "<unknown>";
+      break;
+  }
+  /*
+   * Create the scoped metric
+   * MessageBroker/{Library}/{DestinationType}/{Action}/Named/{DestinationName}
+   * non-temp MessageBroker/{Library}/{DestinationType}/{Action}/Temp
```

Comment:
This comment is confusing - is "non-temp" supposed to be at the end of the previous line?

Reply:
Improved formatting, please see: [12af96f](https://github.com/newrelic/newrelic-php-agent/pull/991/commits/12af96f1c587a461066885a918d0d8cd36d3182c)